### PR TITLE
Issue 5853: Fixed crash in Lua callstack logging

### LINKED ISSUE #5853: Crash in dmScript::GetLuaStackTraceCbk
```
backtrace:
  #00  pc 00000000000859cc  /apex/com.android.runtime/lib64/bionic/libc.so (__memcpy+124)
  #00  pc 00000000000d2f30  /apex/com.android.runtime/lib64/bionic/libc.so (__sfvwrite+148)
  #00  pc 00000000000c9918  /apex/com.android.runtime/lib64/bionic/libc.so (__vfprintf+9784)
  #00  pc 00000000000e61d4  /apex/com.android.runtime/lib64/bionic/libc.so (vsnprintf+184)
  #00  pc 00000000001f510c  REDACTED (dmSnPrintf(char*, unsigned long, char const*, ...)+116)
  #00  pc 0000000000193270  REDACTED (dmScript::GetLuaStackTraceCbk(lua_State*, lua_Debug*, void*)+92)
  #00  pc 000000000018faa0  REDACTED (dmScript::GetLuaTraceback(lua_State*, char const*, void (*)(lua_State*, lua_Debug*, void*), void*)+276)
  #00  pc 00000000001931c4  REDACTED (dmScript::BacktraceErrorHandler(lua_State*)+140)
  #00  pc 00000000001c2b78  REDACTED (lj_BC_FUNCC+40)
  #00  pc 00000000001ab6f8  REDACTED (lj_err_run+404)
  #00  pc 00000000001ab7d8  REDACTED (err_msgv+208)
  #00  pc 00000000001ab7e4  REDACTED (lj_err_msg+8)
  #00  pc 00000000001ae3d0  REDACTED (lj_state_growstack+120)
  #00  pc 00000000001c2d4c  REDACTED (lj_vm_growstack_l+28)
  #00  pc 00000000001b3b64  REDACTED (lua_pcall+124)
  #00  pc 0000000000191b5c  REDACTED (dmScript::PCallInternal(lua_State*, int, int, int)+88)
  #00  pc 000000000023bcdc  REDACTED (dmGui::RunScript(dmGui::Scene*, dmGui::ScriptFunction, int, void*)+2420)
  #00  pc 000000000023d6f8  REDACTED (dmGui::DispatchInput(dmGui::Scene*, dmGui::InputAction const*, unsigned int, bool*)+64)
  #00  pc 0000000000131594  REDACTED (dmGameSystem::CompGuiOnInput(dmGameObject::ComponentOnInputParams const&)+396)
  #00  pc 00000000001051ac  REDACTED (dmGameObject::DispatchInput(dmGameObject::Collection*, dmGameObject::InputAction*, unsigned int)+320)
  #00  pc 0000000000128b78  REDACTED (dmGameSystem::CompCollectionProxyOnInput(dmGameObject::ComponentOnInputParams const&)+44)
  #00  pc 00000000001051ac  REDACTED (dmGameObject::DispatchInput(dmGameObject::Collection*, dmGameObject::InputAction*, unsigned int)+320)
  #00  pc 0000000000128b78  REDACTED (dmGameSystem::CompCollectionProxyOnInput(dmGameObject::ComponentOnInputParams const&)+44)
  #00  pc 00000000001051ac  REDACTED (dmGameObject::DispatchInput(dmGameObject::Collection*, dmGameObject::InputAction*, unsigned int)+320)
  #00  pc 00000000000b3620  REDACTED (dmEngine::Step(dmEngine::Engine*)+744)
  #00  pc 00000000000b46b4  REDACTED (dmEngineUpdate(dmEngine::Engine*)+20)
  #00  pc 00000000000b0a44  REDACTED (dmEngine::RunLoop(dmEngine::RunLoopParams const*)+92)
  #00  pc 00000000000b09c8  REDACTED (EngineMainThread(void*)+124)
  #00  pc 0000000000215ee8  REDACTED (dmThread::ThreadStartProxy(void*)+32)
  #00  pc 00000000000eb7a8  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+64)
  #00  pc 000000000008bc8c  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64)
```

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: engine/engine/src/engine.cpp [hunk 1]
```diff
@@ -457,7 +457,9 @@ namespace dmEngine
 
         if (ctx->m_First)
         {
-            uint32_t nwritten = dmSnPrintf(ctx->m_Buffer, ctx->m_BufferSize, "Lua Callstack:\n");
+            int32_t nwritten = dmSnPrintf(ctx->m_Buffer, ctx->m_BufferSize, "Lua Callstack:\n");
```
Comment (JCash) [inline engine/engine/src/engine.cpp:460]: dmSnPrintf returns -1 if truncated

File: engine/script/src/script.cpp [hunk 2]
```diff
@@ -342,7 +342,12 @@ namespace dmScript
         {
             nwritten = dmSnPrintf(buffer, buffer_size, "  %s:%d: in function <%s:%d>\n", entry->short_src, entry->currentline, entry->short_src, entry->linedefined);
         }
-        return nwritten;
+
+        // Truncated, so we'll just make sure we've not written anything
+        if (nwritten < 0)
+            nwritten = 0;
```
Comment (JCash) [inline engine/script/src/script.cpp:348]: The fix.

Returning 0 means that the next entry will try to write on the same spot (and most likely fail)

File: engine/script/src/test/test_script.cpp [hunk 3]
```diff
@@ -178,6 +180,24 @@ TEST_F(ScriptTest, TestCircularRefPPrint)
...
+    lua_getglobal(L, "do_crash");
+    luaL_checktype(L, -1, LUA_TFUNCTION);
+    int ret = dmScript::PCall(L, 0, LUA_MULTRET);
+
+    ASSERT_EQ(5, ret);
+    ASSERT_EQ(top, lua_gettop(L));
+
+    const char* log = GetLog();
+    printf("LOG: %s\n", log ? log : "");
+}
```
Comment (JCash) [inline engine/script/src/test/test_script.cpp:199]: Running with ASAN catches the error.

**Review by britzl — APPROVED** [review]
Looks good!

**Pull request merged**